fix #7649 Index patterns are not hidden after being deleted

ppisljar opened the pull request:
Closes #7649 - Index patterns are not hidden after being deleted 

added refresh: wait_for to delete index pattern request

Bargs commented:
jenkins, test this

Bargs commented:
@ppisljar I'm still able to reproduce the issue. I don't think the cache clear is the issue, because `ingest.delete()` already calls `clearCache()` https://github.com/elastic/kibana/blob/81efe3b73419acef3c6fc11062511ce4d7e08d60/src/ui/public/ingest/ingest.js#L42-L42

spalger commented:
jenkins, test this

ppisljar commented:
yup, you are right. it worked ok if i have debuger turned on .... so it seems like a race condition. you mentioned you had a similar issue .... any idea what to look at ?

Bargs commented:
This is where the list of index patterns for the left nav gets setup: https://github.com/Bargs/kibana/blob/81efe3b73419acef3c6fc11062511ce4d7e08d60/src/core_plugins/kibana/public/management/sections/indices/index.js#L38-L38

You'll notice that it only updates when the default index pattern changes. I think what I did in the past was set up another listener to update the list when the ["ingest:updated"](https://github.com/Bargs/kibana/blob/81efe3b73419acef3c6fc11062511ce4d7e08d60/src/ui/public/ingest/ingest.js#L43-L43) event is broadcast. For some reason that code change never got in or maybe got wiped out somehow.

ppisljar commented:
this is how i understand it:
- we click on delete index
- we send a request to delete index (ingest)
- ingest will clear indexes cache and broadcast updated event
-- here it seems a race condition occurs ... something else updates the index cache (with old values)
- page will refresh
- on page load indexes will be reloaded (since cache was deleted new values will be loaded ... but something seems to fill in the cache in the meantime with old values)

race condition does not happen if i have debugger attached ... with no debuger it seems to happen everytime.

i was trying to listen to ingest:updated event and call curier string after ... still the race condition happens.

```
      $scope.$on('ingest:updated', function() {
        courier.indexPatterns.getIds().then((ids) => {
          $scope.indexPatternList = ids.map(function (id) {
            return {
              id: id,
              url: kbnUrl.eval('#/management/kibana/indices/{{id}}', {id: id}),
              class: 'sidebar-item-title ' + ($scope.editingId === id ? 'active' : ''),
              default: $scope.defaultIndex === id
            };
          });
        });

      });
```

Bargs commented:
What happens if you wrap the success callback of `getIds` in $scope.$apply?

ppisljar commented:
same thing happens. it really shouldn't matter as the page is reloaded after it right ?

Bargs commented:
The page doesn't get reloaded, it's all xhr

ppisljar commented:
sorry that's not what i meant ... the route gets reloaded (as if you would click on management/index patterns again) .... 

.../sections/indices/_edit.js on line 88:
`$location.url('/management/data/index');`

Bargs commented:
Hmmm, yeah you're right. For some reason the cache just isn't getting cleared properly. Even after the page is fully transitioned, the id cache still has the old values, but if I clear it manually it gets updated:

![screen shot 2016-07-26 at 1 59 20 pm](https://cloud.githubusercontent.com/assets/6239176/17149416/31890896-5339-11e6-9064-fa74a45cbb0f.png)

ppisljar commented:
adding timeout before page load solves the issue ... but i don't like the solution ... block_until_refresh is not helping. do you have any other suggestions ?

ppisljar commented:
jenkins, test this

Bargs commented:
Did you try manually refreshing the kibana index prior to fetching the new ids? https://github.com/Bargs/kibana/blob/81efe3b73419acef3c6fc11062511ce4d7e08d60/src/core_plugins/kibana/public/management/sections/indices/_refresh_kibana_index.js#L1-L1

Bargs commented:
I may have been wrong about the name of `block_until_refresh` when quickly reading through the PR, you could also try out these options https://www.elastic.co/guide/en/elasticsearch/reference/master/docs-refresh.html

ppisljar commented:
great, this solves the issue.

Bargs commented:
@ppisljar I was thinking about this a bit more, and I think we should do the [manual refresh](https://github.com/Bargs/kibana/blob/81efe3b73419acef3c6fc11062511ce4d7e08d60/src/core_plugins/kibana/public/management/sections/indices/_refresh_kibana_index.js#L1-L1) on the client side rather than forcing it on all API requests. It's totally possible that some clients might want to use this delete API endpoint without waiting on the changes to propagate. Sorry for flip flopping, I think I just got excited about the new `wait_for` param.

ppisljar commented:
no problem, i'm learning a lot along the way

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/_edit.js:
```diff
@@ -84,6 +84,7 @@ uiModules.get('apps/management')
     .then(function () {
+      refreshKibanaIndex();
       $location.url('/management/data/index');
     })
```
The race condition still exists in this code, because we're not waiting on the promise from `refreshKibanaIndex` to resolve before changing the route. To do this, we need to change the code to:

```
    ingest.delete($scope.indexPattern.id)
    .then($scope.indexPattern.destroy.bind($scope.indexPattern))
    .then(refreshKibanaIndex)
    .then(function () {
      $location.url('/management/data/index');
    })
```

Bargs commented:
LGTM. Grab another reviewer just to give this a quick sanity check.

ycombinator commented:
Functionality looks good to me. Looking over code now.

ycombinator commented:
LGTM! :shipit:
